### PR TITLE
First cut at providing a state machine.

### PR DESCRIPTION
**Pull request by ekr**
@davidben @martinthomson @FiloSottile 

Here's a first attempt at state machine. I did look at doing the post-handshake messages, but I think it actually destroys information because it's  mostly like "here is a message, now just go back to CONNECTED"

### CODE REVIEW

File: draft-ietf-tls-tls13.md [hunk 1]
```diff
@@ -4375,6 +4375,109 @@ currently deployed and properly configured TLS implementations.
 %%### Ticket Establishment
 %%### Updating Keys
 
+# State Machine
+
+This section provides a summary of the legal state transitions for the
+client and server handshakes. State names have no formal meaning but
+are provided in all capitals (e.g., START) for ease of comprehension.
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: The first several times I read this sentence, I thought you were saying they were in all capitals for ease of comprehension. This was very entertaining but probably not what you meant. :-) Perhaps:

State names (in all capitals, e.g., START) have no formal meaning but are provided for ease of comprehension.

File: draft-ietf-tls-tls13.md [hunk 2]
```diff
@@ -4375,6 +4375,109 @@ currently deployed and properly configured TLS implementations.
...
+       |                 | Recv SH         
+       |                 |                 
+       |                 V                 
+       |              WAIT_EE            
+       |                 |                 
+       |                 | Recv EE         
+       |                 |                 
+       |          +------+------+          
+       |          |             |              
+       |   SH+PSK |             | SH
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: Perhaps:
SH+PSK => Using PSK
SH => Using certificate

My first thought was you sent or received an SH again.

File: draft-ietf-tls-tls13.md [hunk 3]
```diff
@@ -4375,6 +4375,109 @@ currently deployed and properly configured TLS implementations.
...
+         +-----> START
+ Recv CH |      |  |
+ Send HRR|      |  | Recv CH
+         +------+  |
+                   v
+                NEGOTIATED
+                   |
+                   | Send SH
+                   | Send EE
+                   | [Send Cert + CV]
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: CertificateRequest?

File: draft-ietf-tls-tls13.md (hunk 2 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: yeah, I agree. I think when I wrote this I was thinking that we would refer to messages and not to the logic

File: draft-ietf-tls-tls13.md (hunk 3 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Ah, good catch

**Comment by ekr** [comment]
I'll try to remove the abbrevs.

File: draft-ietf-tls-tls13.md [hunk 4]
```diff
@@ -4375,6 +4375,109 @@ currently deployed and properly configured TLS implementations.
...
+       |          +------+------+          
+       |          |             |              
+       |   SH+PSK |             | SH
+       |          |             v
+       |          |         WAIT_CERT_CR
+   Can |          |          |       |
+  Send |          |          |       |
+ Early |          |     Recv |       | Recv CR
+  Data |          |     Cert |       |
+  Here |          |          |       v
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: I'm not sure that you need to highlight the part where early data can be sent.  It's after the first ClientHello, up to the second flight from the client (which might be another ClientHello).  Easier to describe in words and you can reclaim some horizontal space for David's suggestion.

File: draft-ietf-tls-tls13.md [hunk 5]
```diff
@@ -4375,6 +4375,109 @@ currently deployed and properly configured TLS implementations.
...
+                                 v
+                             CONNECTED
+~~~~
+
+## Server
+
+~~~~
+         +-----> START
+ Recv CH |      |  |
+ Send HRR|      |  | Recv CH
```
Comment (ghedo) [inline draft-ietf-tls-tls13.md]: This part seems kind of confusing. If you rename NEGOATIATED to, idk, RECV_CH, then you can make a "Send HRR" transition from there to START without duplicating "Recv CH". You could also keep both RECV_CH and NEGOTIATED and have a "Negotiate" or "Pick parameters" or whatever transition between the two.

```
                 START <--------+
                   |            |
                   | Recv CH    | Send HRR
                   |            |
                   v            |
                RECVED_CH ------+
                   |
                   | Select params
                   |
                   v
                NEGOTIATED
                   |
                   | Send SH
                  ...
```

IMO it makes it clearer (YMMV), with the added bonus that it makes the machine somewhat more deterministic. Con: it adds an additional state.

File: draft-ietf-tls-tls13.md [hunk 6]
```diff
@@ -4375,6 +4375,89 @@ currently deployed and properly configured TLS implementations.
...
+          |             WAIT_CERT
+       No |        Recv |       | Recv Certificate
+     Auth |       Empty |       v
+          | Certificate |    WAIT_CV
+          |             |       | Recv
+          |             |       | CertificateVerify
+          +-> WAIT_FINISHED <---+
+                   | Recv Finished
+                   v
+               CONNECTED
```
Comment (ghedo) [inline draft-ietf-tls-tls13.md]: Should sending and receiving NewSessionTicket also be included?

**Comment by beurdouche** [comment]
@ekr, do you feel that one state machine diagram is enough ? I feel that, except for the main ones, it seems complicated to put all state machines into one diagram, but maybe I am wrong ?!

Might it be valuable for the spec to reference a trusted website with all possible state machine diagrams depending on the negotiated values, extensions... etc... ? ;)

**Comment by beurdouche** [comment]
That said, separating state machines has the major drawback of making it harder for the implementer, so it might be a bad idea too... :( WDYT ?

File: draft-ietf-tls-tls13.md [hunk 7]
```diff
@@ -4375,6 +4375,89 @@ currently deployed and properly configured TLS implementations.
...
+
+## Server
+
+~~~~
+          +----> START  
+     Send |        |
+    Hello |        | Recv ClientHello
+    Retry |        |
+  Request |        v
+          +--- RECVD_CH
```
Comment (mimoo) [inline draft-ietf-tls-tls13.md]: What is this "RECVD" ? Receive Daemon?

**Comment by hannesm** [comment]
I do appreciate to have ASCII art statemachines in the RFC (and not referencing some random other resource on the internet).  About the concrete proposal, should it be the case that arrows represent actions (recv message/send message/choose & verify parameters), and states are either start or established or WAIT_FOR_XXX (if the server has to send multiple messages, coalesce them into one arrow -- or should each action have its own arrow, and there are intermediate state (like certificate_request_sent_next_sending_certificate))?

File: draft-ietf-tls-tls13.md [hunk 8]
```diff
@@ -4375,6 +4375,89 @@ currently deployed and properly configured TLS implementations.
...
+          |             |    WAIT_CERT
+          |             |       | Recv Certificate
+          |             v       v
+          |              WAIT_CV
+          |                 | Recv CertificateVerify
+          +> WAIT_FINISHED <+
+                   | Recv Finished
+                   |
+                   | [Send EndOfEarlyData]
+                   | [Send Certificate [+ CertificateVerify]]
```
Comment (mimoo) [inline draft-ietf-tls-tls13.md]: Why not `[Send Certificate + Certificate Verify]` instead since these messages must come together?

Comment (mimoo) [inline draft-ietf-tls-tls13.md]: (This is actually what you do on the server side.)

Comment (ekr) [inline draft-ietf-tls-tls13.md]: The way that the client rejects client auth is by sending an empty certificate message.

File: draft-ietf-tls-tls13.md [hunk 9]
```diff
@@ -4375,6 +4375,89 @@ currently deployed and properly configured TLS implementations.
...
+                NEGOTIATED
+                   | Send ServerHello
+                   | Send EncryptedExtensions
+                   | [Send CertificateRequest]
+                   | [Send Certificate + CertificateVerify]
+                   | Send Finished
+          +--------+--------+
+ No 0-RTT |                 | 0-RTT
+          |                 v
+          |             WAIT_EOED <---+
```
Comment (mimoo) [inline draft-ietf-tls-tls13.md]: It's weird that this message (EndOfEarlyData) is written unabbreviated then abbreviated. I know that it takes a lot of room though :)

File: draft-ietf-tls-tls13.md [hunk 10]
```diff
@@ -4375,6 +4375,89 @@ currently deployed and properly configured TLS implementations.
...
+          |        Recv |       |     | Recv
+          |        EOED |       |     | Application Data
```
Comment (mimoo) [inline draft-ietf-tls-tls13.md]: It's weird that this message (EndOfEarlyData) is written unabbreviated then abbreviated. I know that it takes a lot of room though :)

**Comment by beurdouche** [comment]
@hannesm, sure, the spec must be self contained.
The question is, what do you do with future extensions that branch the state-machine. Once finalized the spec we won't be able to reference external ressources with up-to-date state-machines. 
So the point is, keep the main stuff here but should we eventually also point to something up-to-date on the internet.

File: draft-ietf-tls-tls13.md [hunk 11]
```diff
@@ -4375,6 +4375,89 @@ currently deployed and properly configured TLS implementations.
...
+          |             |       +-----+
+          +> WAIT_FLT2 <+
```
Comment (mimoo) [inline draft-ietf-tls-tls13.md]: What is this FLT2?

File: draft-ietf-tls-tls13.md (hunk 8 above)
Comment (mimoo) [inline draft-ietf-tls-tls13.md]: ah right!

File: draft-ietf-tls-tls13.md [hunk 12]
```diff
@@ -4375,6 +4375,89 @@ currently deployed and properly configured TLS implementations.
...
+  Request |        v
+          +--- RECVD_CH
+                   | Select parameters
+                   v
+                NEGOTIATED
+                   | Send ServerHello
+                   | Send EncryptedExtensions
+                   | [Send CertificateRequest]
+                   | [Send Certificate + CertificateVerify]
+                   | Send Finished
```
Comment (mimoo) [inline draft-ietf-tls-tls13.md]: might want to add the "early data" of the server as well
```
| Send Finished
| [Send ApplicationData]
```

**Comment by ekr** [comment]
On Mon, Dec 19, 2016 at 9:24 AM, David Wong <notifications@github.com>
wrote:

> *@mimoo* commented on this pull request.
> ------------------------------
>
> In draft-ietf-tls-tls13.md
> <https://github.com/tlswg/tls13-spec/pull/824#pullrequestreview-13599802>:
>
> > @@ -4375,6 +4375,89 @@ currently deployed and properly configured TLS implementations.
>  %%### Ticket Establishment
>  %%### Updating Keys
>
> +# State Machine
> +
> +This section provides a summary of the legal state transitions for the
> +client and server handshakes.  State names (in all capitals, e.g.,
> +START) have no formal meaning but are provided for ease of
> +comprehension.  Messages which are sent only sometimes are indicated
> +in [].
> +
> +## Client
> +
> +~~~~
> +          +-----> START
> +     Recv |        |
> +    Hello |        | Send ClientHello
>
> might want to include early data in here:
>
> send ClientHello [+ application_data] (early_data)
>

MT talked me out of it. Said it was better in text.

**Comment by mimoo** [comment]
Mmm, I think it is fair to assume that some implementations might want to include a whitelist of state transitions, and that they will probably look at these diagrams to draft it from (if it is not included in the final document). It's also a pretty small addition for a lot of meaning :) (especially for the server "early data", I don't think it is detailed anywhere besides the [overview section](https://tlswg.github.io/tls13-spec/#incorrect-dhe-share)

File: draft-ietf-tls-tls13.md (hunk 10 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: yeah, I am leaving the states abbreviated.

File: draft-ietf-tls-tls13.md (hunk 11 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Flight 2. It's just saving room.

File: draft-ietf-tls-tls13.md (hunk 7 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: "I have received"

**Comment by ekr** [comment]
@beurdouche: perhaps the place to have an up-to-date diagram is on the wiki in github?

**Comment by ekr** [comment]
@hannesm I think I prefer the current style to having extra states that don't represent choice points

**Comment by ekr** [comment]
I just updated with markers to show when you could send app data. Comments welcome

File: draft-ietf-tls-tls13.md [hunk 13]
```diff
@@ -4375,6 +4375,91 @@ currently deployed and properly configured TLS implementations.
...
+in [].
+
+## Client
+
+~~~~
+                  +-----> START
+             Recv |        |
+            Hello |        | Send ClientHello
+            Retry |        v
+        +-> Request +---- WAIT_SH
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: Something about the alignment here is a bit off. I think you lost two spaces on the first few lines here:

```
                    +-----> START
               Recv |        |
              Hello |        | Send ClientHello
              Retry |        v
        +-> Request +---- WAIT_SH
    Can |                    | Recv ServerHello
   Send |                    V
```

Hyphens might also be useful:

```
                    +-----> START
               Recv |        |
             Hello- |        | Send ClientHello
             Retry- |        v
        +-> Request +---- WAIT_SH
    Can |                    | Recv ServerHello
   Send |                    V
```

Or you can then avoid line-wrapping if you swap the branches:

```
                            START <---+
                             |        |
            Send ClientHello |        | Recv HelloRetryRequest
                             v        |
        +->               WAIT_SH ----+
    Can |                    | Recv ServerHello
   Send |                    V
```

File: draft-ietf-tls-tls13.md [hunk 14]
```diff
@@ -4375,6 +4375,91 @@ currently deployed and properly configured TLS implementations.
...
+~~~~
+
+## Server
+
+~~~~
+                      +----> START
+                 Send |        |
+                Hello |        | Recv ClientHello
+                Retry |        |
+              Request |        v
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: If you swap the order, I think you have room to leave them both unwrapped here too:

```
                             START <----+
                               |        |
              Recv ClientHello |        | Send HelloRetryRequest
                               |        |
                               v        |
                           RECVD_CH ----+
                               | Select parameters
                               v
                            NEGOTIATED
                               | Send ServerHello
```

File: draft-ietf-tls-tls13.md [hunk 15]
```diff
@@ -4375,6 +4375,91 @@ currently deployed and properly configured TLS implementations.
...
+        |           | Certificate |       v
+        |           |             |    WAIT_CERT
+        |           |             |       | Recv Certificate
+        |           |             v       v
+        |           |              WAIT_CV
+        |           |                 | Recv CertificateVerify
+        |           +> WAIT_FINISHED <+
+        |                  | Recv Finished
+        |                  |
+        +->                | [Send EndOfEarlyData]
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: This should end just before EndOfEarlyData, right?

What do you think about changing the ends to not be arrows? If you keep HRR on the left side, it looks kinda like a state transition. How about?

```
                            START <---+
                             |        |
            Send ClientHello |        | Recv HelloRetryRequest
        /                    v        |
        |                 WAIT_SH ----+
    Can |                    | Recv ServerHello
   Send |                    V
  Early |                 WAIT_EE
   Data |                    | Recv EncryptedExtensions
        |           +--------+--------+
        |     Using |                 | Using Certificate
        |       PSK |                 v
        |           |            WAIT_CERT_CR
        |           |        Recv |       | Recv CertificateRequest
        |           | Certificate |       v
        |           |             |    WAIT_CERT
        |           |             |       | Recv Certificate
        |           |             v       v
        |           |              WAIT_CV
        |           |                 | Recv CertificateVerify
        |           +> WAIT_FINISHED <+
        |                  | Recv Finished
        \                  |
                           | [Send EndOfEarlyData]
                           | [Send Certificate [+ CertificateVerify]]
                           | Send Finished
 Can Send                  v
 App Data -->          CONNECTED
 After
 Here
```

File: draft-ietf-tls-tls13.md [hunk 16]
```diff
@@ -4375,6 +4375,91 @@ currently deployed and properly configured TLS implementations.
...
+                               | Select parameters
+                               v
+                            NEGOTIATED
+                               | Send ServerHello
+                               | Send EncryptedExtensions
+                               | [Send CertificateRequest]
+Can Send                       | [Send Certificate + CertificateVerify]
+App Data -->                   | Send Finished
+After                 +--------+--------+
+Here            0-RTT |                 | 0-RTT
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: Left 0-RTT should be "No 0-RTT"?

File: draft-ietf-tls-tls13.md [hunk 17]
```diff
@@ -4375,6 +4375,91 @@ currently deployed and properly configured TLS implementations.
...
+                      |                 v
+                      |             WAIT_EOED <---+
+                      |        Recv |       |     | Recv
+                      |        EOED |       |     | Early Data
+                      |             |       +-----+
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: EOED is still abbreviated. I feel like ideally we'd also have that be the "core" branch while the WAIT_EOED self-loop is more a side thing. That also solves our abbreviation problem:

```
Can Send                       | [Send Certificate + CertificateVerify]
App Data -->                   | Send Finished
After                 +--------+--------+
Here         No 0-RTT |                 | 0-RTT
                      |                 v
                      |             WAIT_EOED ---+
                      |            Recv |  ^     | Recv Early Data
                      |  EndOfEarlyData |  |     | 
                      |                 |  +-----+
                      +--> WAIT_FLT2 <--+
                               |
```

Comment (davidben) [inline draft-ietf-tls-tls13.md]: Oh, I'd maybe move the "Can Send App Data After Here" down one line just so it's clear it's after "Send Finished" rather before.

File: draft-ietf-tls-tls13.md [hunk 18]
```diff
@@ -4375,6 +4375,91 @@ currently deployed and properly configured TLS implementations.
...
+## Server
+
+~~~~
+                      +----> START
+                 Send |        |
+                Hello |        | Recv ClientHello
+                Retry |        |
+              Request |        v
+                      +--- RECVD_CH
+                               | Select parameters
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: Capitalization is also a bit inconsistent. For the most part, Everything Is Written In Title Case:
- Can Send Early Data
- Can Send App Data After Here
- Client Auth
- No Auth
- Using Certificate
- Recv Early Data
- Recv Empty Certificate

But "Select parameters" is sentence-case. FWIW, I think the Excessive Title Case Feels A Little Weird and might suggest changing the others instead. (Though keep messages in UpperCamelCase of course.) These feel more like sentences.

File: draft-ietf-tls-tls13.md [hunk 19]
```diff
@@ -4375,6 +4375,91 @@ currently deployed and properly configured TLS implementations.
...
+                      |                 v
+                      |             WAIT_EOED <---+
+                      |        Recv |       |     | Recv
+                      |        EOED |       |     | Early Data
+                      |             |       +-----+
+                      +> WAIT_FLT2 <+
+                               |
+                      +--------+--------+
+                      |                 | Client
+                      |                 | Auth
```
Comment (davidben) [inline draft-ietf-tls-tls13.md]: I think you can fit "Client Auth" and "No Auth" on one line each.

Labels on state transitions are also inconsistent about whether they're vertically-aligned at top or center. "No Auth" is center, "Recv Empty Certificate" is top, and "Client Auth" is either top or a rounding error. (Maybe standardize on top so you don't have rounding errors? Top seems more common.)

Comment (ekr) [inline draft-ietf-tls-tls13.md]: The reason I did top versus center is that there were places where the labels were too short to accommodate the amount of horizontal spacing needed to separate the two branches, as with WAIT_FLT2. I'll see about making them longer and maybe that will help

**Pull request merged**